### PR TITLE
Compile with C++23

### PR DESCRIPTION
**Pull request by rschu1ze**
Enables compilation with C++23 (at least the [subset](https://en.cppreference.com/w/cpp/23) of it that Clang 15 supports).

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
ClickHouse now builds with C++23

### CODE REVIEW

File: base/base/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,10 @@ if (USE_CLANG_TIDY)
     set (CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_PATH}")
 endif ()
 
+# TODO: Remove this. We generally like to compile with C++23 but Clang currently crashes with our libcxx
```
Comment (qoega) [inline base/base/CMakeLists.txt:5]: Did we report it to clang? And what about clang-16?

Comment (rschu1ze) [reply base/base/CMakeLists.txt:5]: Will try clang 16 once it's released. I just know that @amosbird recently did some fixes in preparation for clang 16 (#47027).

I would assume that just a newer compiler won't help. We'd likely also need to upgrade libcxx and libcxxabi (+ our internal LLVM).

Perhaps it's better then to defer this PR by a little bit and make it dependent on Clang 16.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by rschu1ze** [comment]
clang-tidy became terribly slow ... let me try to disable a few checks.

**Comment by rschu1ze** [comment]
Turned out after some tedious bisecting that clang-tidy checks `readability-identifier-naming` and `bugprone-reserved-identifier` (and its aliases) became for some reason super slow when C++23 is enabled. Disabling them.

**Comment by alexey-milovidov** [comment]
Could you please file a bug report to LLVM?

**Comment by rschu1ze** [comment]
Here we go: https://github.com/llvm/llvm-project/issues/61418

**Pull request merged**